[Black Jack]: Comparisons Concept and Concept Exercise Finalization

BethanyG opened the pull request:
Added/Changed/Edited files:

- [x]   config.json
- [x]   exercises/concept/black-jack/.meta/config.json
----
- [x]   concepts/comparisons/introduction.md
- [x]   concepts/comparisons/about.md
____
- [x]   exercises/concept/black-jack/.docs/introduction.md
- [x]   exercises/concept/black-jack/.docs/instructions.md
- [x]   exercises/concept/black-jack/.docs/hints.md
____
- [x]  exercises/concept/black-jack/.meta/exemplar.py
- [x]  exercises/concept/black-jack/black_jack.py
- [x]   exercises/concept/black-jack/black_jack_test.py

kotp commented on concepts/comparisons/about.md:
```diff
@@ -1,199 +1,291 @@
+| `<=`     | "less than or equal to"    | `a <= b` is `True` if `a < b` or `a == b` in value             |
+| `!=`     | "not equal to"             | `a != b` is `True` if `a == b` is `False`                      |
+| `is`     | "identity"                 | `a is b` is `True` if **_and only if_** `a` and `b` are the same _object_  |
+| `is not` | "negated identity"                 | `a is not b` is `True` if `a` and `b` are **not** the same _object_ |
```
If we are alligning for the table we should do so.

BethanyG replied:
There's a big difference in the way this appears in code vs render.  When rendered, this is aligned: 

| Operator | Operation                | Description                                                    |
| -------- | ------------------------ | -------------------------------------------------------------- |
| `>`      | "greater than"             | `a > b` is `True` if `a` is **strictly** greater in value than `b` |
| `<`      | "less than"                | `a < b` is `True` if `a` is **strictly** less in value than `b`    |
| `==`     | "equal to"                 | `a == b` is `True` if `a` is **strictly** equal to `b` in value    |
| `>=`     | "greater than or equal to" | `a >= b` is `True` if `a > b` OR `a == b` in value             |
| `<=`     | "less than or equal to"    | `a <= b` is `True` if `a < b` or `a == b` in value             |
| `!=`     | "not equal to"             | `a != b` is `True` if `a == b` is `False`                      |
| `is`     | "identity"                 | `a is b` is `True` if **_and only if_** `a` and `b` are the same _object_  |
| `is not` | "negated identity"                 | `a is not b` is `True` if `a` and `b` are **not** the same _object_ |
| `in`     | "containment test"         | `a in b` is `True` if `a` is member, subset, or element of `b`               |
| `not in` | "negated containment test" | `a not in b` is `True` if `a` is not a member, subset, or element of `b`     |

BethanyG replied:
Neverimind.  Prettier to the rescue!

kotp replied:
Yep, I knew that it would look good rendered (markdown is awesome, tables are cool that way), but having it look good in source is nice too.

The other choice is "less white space" and not align any of it, but have it naturally end, which I do not _not_ like choice as well.

But having it inconsistent, well, that is distracting. ;)

kotp commented on concepts/comparisons/about.md:
If the `<>` for this is meant to be code, we should do something similar as done in the "and only if" statement where it is wrapped in code delimiters inline.

BethanyG replied:
Yeah. I debated that.  This way, it emphasizes the `is` operator, but requires I escape the `<`.  Not sure I like how `<apple> is <orange>` reads as one block.  We _could_ do `<apple>` `is` `<orange>` ...dunno if that is better.  Thoughts?

kotp replied:
I think that I notice the delimiters  and generally ignore the name of the variables being used anyway.  So to me the `<>` is distracting from the communication, while if I saw `apple is orange` where the reading is talking about `is` and `is not` I would see the example as intended, that the names on either side do not really matter, the "how to use the identity operator `is` or `is not`.  It is why we see `foo` and `bar` as nonsense names in code documentation, we learn that those are not the important part, only that something is there in each place.

kotp commented on concepts/comparisons/about.md:
It looks like the code delimiter changed here as well, perhaps unintended.

BethanyG replied:
Same as above -- not sure what reads better, but a whole block doesn't read well to me.

kotp commented on concepts/comparisons/introduction.md:
```diff
@@ -1,23 +1,65 @@
+| `<=`     | "less than or equal to"    | `a <= b` is `True` if `a < b` or `a == b` in value             |
+| `!=`     | "not equal to"             | `a != b` is `True` if `a == b` is `False`                      |
+| `is`     | "identity"                 | `a is b` is `True` if **_and only if_** `a` and `b` are the same _object_  |
+| `is not` | "negated identity"                 | `a is not b` is `True` if `a` and `b` are **not** the same _object_ |
```
Same here for alignment of column delimiter.

BethanyG replied:
Same comment as above.  The render is more important than the code -- and the render is in alignment.

kotp commented on concepts/comparisons/introduction.md:
```diff
@@ -1,23 +1,65 @@
-[arithmetic conversions]: https://docs.python.org/3/reference/expressions.html?highlight=number%20conversion#arithmetic-conversions
+They all have the same priority (_which is higher than that of [Boolean operations][boolean operations], but lower than that of arithmetic or bitwise operations_).
+
+
```
Two blank spaces which is usually ignored in HTML rendering, and not really helpful in the plain text viewing.  This happens in other places as well.  Single blank line between parenthesis is likely intended.

BethanyG replied:
I am not sure I follow here.  Which two spaces are we talking about?

kotp replied:
Two blank vertical spaces (new lines) is what I really mean here.

kotp commented on exercises/concept/black-jack/.docs/introduction.md:
```diff
@@ -1,185 +1,249 @@
+| `<=`     | "less than or equal to"    | `a <= b` is `True` if `a < b` or `a == b` in value             |
+| `!=`     | "not equal to"             | `a != b` is `True` if `a == b` is `False`                      |
+| `is`     | "identity"                 | `a is b` is `True` if **_and only if_** `a` and `b` are the same _object_  |
+| `is not` | "negated identity"                 | `a is not b` is `True` if `a` and `b` are **not** the same _object_ |
```
Another column delimiter off kilter.

kotp commented on exercises/concept/black-jack/.docs/introduction.md:
```diff
@@ -1,185 +1,249 @@
+The operators `is` and `is not` test for object [_identity_][object identity], as opposed to object _value_.
+An object's identity never changes after creation and can be found by using the [`id()`][id function] function.
+
+\<apple\> `is` \<orange\> evaluates to `True` if _**and only if**_ `id(<apple>)` == `id(<orange>)`.
```
Starting to feel like this is coming from some place higher up the chain maybe, as this seems like it should be

    apple is orange

We can compare apples to oranges, and those would be valid names, but not sure what the `<>` does to them, though I think the communication that I get from reading about apples and oranges are a generation, kind of like `foo` or `bar`, and maybe the `<foo>` and `<bar>` does something different in Python specifically.

So maybe take the feedback like you might with apple slices, you know, with a little salt.

BethanyG replied:
😆   So..same comment here as I had earlier.  We just need to decide which reads better.  The _intent_ of the `<` is to denote something similar to `<foo>` and `<bar>` - placeholders -- just not `foo` and `bar` placeholders.

kotp replied:
Right, but in this case, I see the `<>` and that yells pay attention to this, because that is what delimiters do for me in code.

kotp commented on exercises/concept/black-jack/.docs/introduction.md:
```diff
@@ -1,185 +1,249 @@
+An object's identity never changes after creation and can be found by using the [`id()`][id function] function.
+
+\<apple\> `is` \<orange\> evaluates to `True` if _**and only if**_ `id(<apple>)` == `id(<orange>)`.
+apple `is not` orange yields the inverse.
```
I think this should be 

```markdown
`apple is not orange` yields the inverse.
```

instead.

BethanyG replied:
yeah..or whatever we end up deciding.  Although it should be `<apple> is not <orange>` -- if we want to keep to the `<>` signifying a variable or object.

kotp replied:
If that `<>` indicates variable or object in Python speak then that is what should be used.  I think in the context, perhaps it is clear to focus on the `is not` rather than the variable name used (and is why we use foo and bar as well known "can be whatever, as allowed by the syntax of the language in that case sensitive interpretation".

kotp commented on exercises/concept/black-jack/.docs/introduction.md:
```diff
@@ -1,185 +1,249 @@
+\<apple\> `is` \<orange\> evaluates to `True` if _**and only if**_ `id(<apple>)` == `id(<orange>)`.
+apple `is not` orange yields the inverse.
+
+Due to their singleton status, `None` and `NotImplemented` should always be compared to items using `is` and `is not`.
```
Maybe expand on why, other than only "because of their singleton status".  I am sure there is something more implied here.  And I can not think of why I would ever do other comparisons, anyway.  So the fact that it is mentioned and specifically mentioned that they are singletons, raises an eyebrow, wanting more information.

BethanyG replied:
I can put a link into the python docs.  We're practically quoting them here.  The `singleton status` -- is exactly that -- there is only ever ONE instance of either `None` or `NotImplemented` in any running Python program.  So equivalence is never the right metaphor.....it is either **THE** `None` or **THE** `NotImplemented` - nothing is ever equal to them, because nothing can ever BE them .... except themselves.

BethanyG replied:
Here is the relevant snip:

BethanyG replied:
..I guess we could also link to PEP8 - but the length of that document might confuse someone who is trying to just learn about comparisons.

kotp replied:
I like that, I think the link to that section is good.  I like the addition of "never the equality operators".  A little more prodding to the right idea here.  That we check for identity not equality when asking if `ourThing is None`, or if `ourBehavior is not NotI'mplemented` if that is correct.

kotp commented on exercises/concept/black-jack/.meta/exemplar.py:
```diff
@@ -1,68 +1,93 @@
+How to play blackjack:    https://bicyclecards.com/how-to-play/blackjack/
+"Standard" playing cards: https://en.wikipedia.org/wiki/Standard_52-card_deck
+
+"""
```
Why a blank line around this string, and not the following?

BethanyG replied:
[PEP257](https://www.python.org/dev/peps/pep-0257/) puts the closing quotes on their own lines.  PyCharm defaults to inserting and extra line above - and since I think it also reads better that way, I left it.

BethanyG replied:
I'll normalize it.

kotp commented on exercises/concept/black-jack/black_jack.py:
```diff
@@ -1,50 +1,64 @@
+How to play blackjack:    https://bicyclecards.com/how-to-play/blackjack/
+"Standard" playing cards: https://en.wikipedia.org/wiki/Standard_52-card_deck
+
+"""
```
Also blank line around the string content unlike some other `"""` strings.

kotp reviewed (changes requested):
Looking pretty good, but I think there may have been some unintended changes, plus what appears to be style deviance instances in the documents themselves.

And a question about `<apples>` and `<oranges>`.

BethanyG commented:
Alrighty.  Ran everything through priettier, and cleaned up the `<apples>` and `<oranges>` 😉   I also removed the excess lines in the module docstrings.  Let me know if there's anything I've missed.  😄   Thanks for reviewing this!

BethanyG commented:
@kotp - are there additional changes needed here for approval?  Let me know - thanks!

kotp commented:
I approve, as long as `<fish>` is valid Python.  I kind of think it may be confusing if the student plus that in and it does not work.  And then has to interpret it more as Unix documentation style for the meanings of those brackets.

BethanyG commented:
I have no bead on if `<fish>` is going to be confusing.  The convention is used in many Python concept exercises to denote a placeholder variable or parameter, to signal to the student that the _name_ can really be anything they want.

My thought is to have the exercise go live in `beta` mode, and if it is confusing to students they will (loudly) let us know - and we can discuss re-working it at that point.

kotp commented:
On 22/01/08 10:27AM, BethanyG wrote:
> I have no bead on if `<fish>` is going to be confusing.  The
> convention is used in many Python concept exercises to denote a
> placeholder variable or parameter, to signal to the student that the
> _name_ can really be anything they want.

Also, if it has been being used, and understood well,
already, then I do not have a problem with it.

> My thought is to have the exercise go live in `beta` mode, and if it
> is confusing to students they will (loudly) let us know - and we can
> discuss re-working it at that point.

Definitely.  I do not know if this is common outside of the Python
track, and used widely in the Python greater community.  As an
"outsider" I definitely that as a perspective.  But if it is convention
in the Python community, then it should be used here as well, as the
student will learn that this is not something unique in how this is
communicated on Exercism.
